services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/QueueAttributesManager.java

```diff
@@ -0,0 +1,181 @@
+ *
+ * <p>This class uses an {@link AtomicReference} to maintain the state of the attribute map, allowing concurrent access
+ * and handling cases where the fetching of attributes may fail. If an error occurs during the retrieval of attributes,
+ * the state is reset to allow for a fresh attempt in subsequent calls.</p>
+ *
+ * <p>The class provides methods to get the visibility timeout and calculate the message receive timeout, which
+ * are asynchronously retrieved and processed using {@link CompletableFuture}. These methods handle cancellation
+ * scenarios by cancelling the SQS request if the calling future is cancelled.</p>
+ *
+ * <p>This class is intended for internal use and is marked with the {@link SdkInternalApi} annotation.</p>
+ */
+@SdkInternalApi
+public final class QueueAttributesManager {
+
+    private static final List<QueueAttributeName> QUEUE_ATTRIBUTE_NAMES =
+        Arrays.asList(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS,
+                      QueueAttributeName.VISIBILITY_TIMEOUT);
+    private final SqsAsyncClient sqsClient;
+    private final String queueUrl;
+    private final AtomicReference<CompletableFuture<Map<QueueAttributeName, String>>> queueAttributeMap = new AtomicReference<>();
+
+    public QueueAttributesManager(SqsAsyncClient sqsClient, String queueUrl) {
+        this.sqsClient = sqsClient;
+        this.queueUrl = queueUrl;
+    }
+
+    /**
+     * Retrieves the received message timeout based on the provided request and queue attributes.
+     *
+     * @param rq                 The receive message request
+     * @param configuredWaitTime The configured minimum wait time
+     * @return CompletableFuture with the calculated receive message timeout in milliseconds
+     */
+    public CompletableFuture<Duration> getReceiveMessageTimeout(ReceiveMessageRequest rq, Duration configuredWaitTime) {
+        Integer waitTimeSeconds = rq.waitTimeSeconds();
+        if (waitTimeSeconds != null) {
+            long waitTimeMillis = TimeUnit.SECONDS.toMillis(waitTimeSeconds);
+            return CompletableFuture.completedFuture(Duration.ofMillis(Math.max(configuredWaitTime.toMillis(), waitTimeMillis)));
+        }
+
+        CompletableFuture<Map<QueueAttributeName, String>> attributeFuture = getAttributeMap();
+        CompletableFuture<Duration> resultFuture = attributeFuture.thenApply(attributes -> {
+            String waitTimeSecondsStr = attributes.get(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS);
+            long waitTimeFromSqsMillis = TimeUnit.SECONDS.toMillis(Long.parseLong(waitTimeSecondsStr));
+            return Duration.ofMillis(Math.max(configuredWaitTime.toMillis(), waitTimeFromSqsMillis));
+        });
+
+        resultFuture.whenComplete((r, t) -> {
+            if (t instanceof CancellationException) {
+                attributeFuture.cancel(true);
+            }
+        });
```

Comment:
We can use https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/CompletableFutureUtils.java#L75